README.md
[circleci-image]: https://circleci.com/gh/n3r4zzurr0/range-slider-input.svg?style=shield
[circleci-url]: https://circleci.com/gh/n3r4zzurr0/range-slider-input
[npm-image]: https://img.shields.io/npm/v/range-slider-input.svg
[npm-url]: https://npmjs.org/package/range-slider-input
[size-image]: https://img.shields.io/bundlephobia/minzip/range-slider-input@latest
[size-url]: https://bundlephobia.com/result?p=range-slider-input@latest
[vulnerabilities-image]: https://snyk.io/test/npm/range-slider-input/badge.svg
[vulnerabilities-url]: https://snyk.io/test/npm/range-slider-input
[standard-image]: https://img.shields.io/badge/code_style-standard-brightgreen.svg
[standard-url]: https://standardjs.com
[license-image]: https://img.shields.io/github/license/n3r4zzurr0/range-slider-input.svg
[license-url]: https://github.com/n3r4zzurr0/range-slider-input/blob/main/LICENSE

# range-slider-input
[![circleci][circleci-image]][circleci-url] [![npm][npm-image]][npm-url] [![minzipped size][size-image]][size-url] [![known vulnerabilities][vulnerabilities-image]][vulnerabilities-url] [![javascript style guide][standard-image]][standard-url] [![license][license-image]][license-url]

A lightweight (~2kB) library to create range sliders that can capture a value or a range of values with one or two drag handles.

**[Examples](https://n3r4zzurr0.in/range-slider-input/) / [CodePen](https://codepen.io/n3r4zzurr0/pen/eYVJBMV)**

[![Demo](https://n3r4zzurr0.in/static/rsi-demo600.gif)](https://n3r4zzurr0.in/range-slider-input/)

:sparkles: **Features**
- High CSS customizability
- Touch and keyboard accessible
- Supports negative values
- Vertical orientation
- Small and fast
- Zero dependencies
- Supported by all major browsers
- Has a [React component wrapper](#component-wrappers)

<hr>

:warning: **It is recommended that you upgrade from v1.x to v2.x! [What's new and what's changed in v2.x?](https://github.com/n3r4zzurr0/range-slider-input/blob/main/CHANGELOG.md#v20x)**

<hr>
<br>


## Installation

**npm**

```
npm install range-slider-input
```

Import the `rangeSlider` constructor and the core CSS:

```js
import rangeSlider from 'range-slider-input';
import 'range-slider-input/dist/style.css';
```

**CDN**

```html
<script src="https://cdn.jsdelivr.net/npm/range-slider-input@2.4/dist/rangeslider.umd.min.js"></script>
```

or

```html
<script src="https://unpkg.com/range-slider-input@2"></script>
```

The core CSS comes bundled with the jsDelivr and unpkg imports.


## Usage
```js
import rangeSlider from 'range-slider-input';
import 'range-slider-input/dist/style.css';

const rangeSliderElement = rangeSlider(element);
```


## API

### `rangeSlider(element, options = {})`

Returns an object of functions that can be called to read or write the properties initially set by `options`.

### Parameters

#### `element`

[`HTMLElement`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement)

#### `options`

Object that specifies the characteristics of the range slider element with the following available properties:

<table>
<tr>
    <th>Property</th>
    <th>Type</th>
    <th>Default value</th>
    <th>Description</th>
</tr>
<tr>
    <td><code>min</code></td>
    <td>number</td>
    <td>0</td>
    <td>Number that specifies the lowest value in the range of permitted values.<br>Its value must be less than that of <code>max</code>.</td>
</tr>
<tr>
    <td><code>max</code></td>
    <td>number</td>
    <td>100</td>
    <td>Number that specifies the greatest value in the range of permitted values.<br>Its value must be greater than that of <code>min</code>.</td>
</tr>
<tr>
    <td><code>step</code></td>
    <td>number / string</td>
    <td>1</td>
    <td>Number that specifies the amount by which the slider value(s) will change upon user interaction.<br>Other than numbers, the value of <code>step</code> can be a string value of <code>any</code>.<br><br>From <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#step">MDN</a>,<blockquote> A string value of <code>any</code> means that no stepping is implied, and any value is allowed (barring other constraints, such as min and max).</blockquote></td>
</tr>
<tr>
    <td><code>value</code></td>
    <td>number[]</td>
    <td>[25, 75]</td>
    <td>Array of two numbers that specify the values of the lower and upper offsets of the range slider element respectively.</td>
</tr>
<tr>
    <td><code>onInput</code></td>
    <td>function</td>
    <td>NOOP</td>
    <td>Function to be called when there is a change in the value(s) of range sliders upon user interaction or upon calling <a href="#min-max-step-value-and-orientation"><code>min()</code></a>, <a href="#min-max-step-value-and-orientation"><code>max()</code></a>, <a href="#min-max-step-value-and-orientation"><code>step()</code></a>, <a href="#min-max-step-value-and-orientation"><code>value()</code></a> or <a href="#min-max-step-value-and-orientation"><code>orientation()</code></a>.<br><br><b>Usage:</b> <code>(value, userInteraction) => {}</code><br><br><code>value</code> holds the current lower and upper values in an array and <code>userInteraction</code> is a boolean value which is <code>true</code> if the value is changed upon user interaction.</td>
</tr>
<tr>
    <td><code>onThumbDragStart</code></td>
    <td>function</td>
    <td>NOOP</td>
    <td>Function to be called when the <code>pointerdown</code> event is triggered for any of the thumbs.</td>
</tr>
<tr>
    <td><code>onThumbDragEnd</code></td>
    <td>function</td>
    <td>NOOP</td>
    <td>Function to be called when the <code>pointerup</code> event is triggered for any of the thumbs.</td>
</tr>
<tr>
    <td><code>onRangeDragStart</code></td>
    <td>function</td>
    <td>NOOP</td>
    <td>Function to be called when the <code>pointerdown</code> event is triggered for the range.</td>
</tr>
<tr>
    <td><code>onRangeDragEnd</code></td>
    <td>function</td>
    <td>NOOP</td>
    <td>Function to be called when the <code>pointerup</code> event is triggered for the range.</td>
</tr>
<tr>
    <td><code>disabled</code></td>
    <td>boolean</td>
    <td>false</td>
    <td>Boolean that specifies if the range slider element is disabled or not.</td>
</tr>
<tr>
    <td><code>rangeSlideDisabled</code></td>
    <td>boolean</td>
    <td>false</td>
    <td>Boolean that specifies if the range is slidable or not.</td>
</tr>
<tr>
    <td><code>thumbsDisabled</code></td>
    <td>boolean[]</td>
    <td>[false, false]</td>
    <td>Array of two Booleans which specify if the lower and upper thumbs are disabled or not, respectively. If only one Boolean value is passed instead of an array, the value will apply to both thumbs.</td>
</tr>
<tr>
    <td><code>orientation</code></td>
    <td>string</td>
    <td>horizontal</td>
    <td>String that specifies the axis along which the user interaction is to be registered. By default, the range slider element registers the user interaction along the X-axis. It takes two different values: <code>horizontal</code> and <code>vertical</code>.</td>
</tr>
</table>

### Return value

Object of functions that can be called to read or write the properties initially set by the `options` parameter. Available functions:

#### `min()`, `max()`, `step()`, `value()` and `orientation()`

These are simple getter and setter functions. So, while calling these functions, if a parameter is supplied, the corresponding values will be set, and if a parameter is not supplied, the corresponding values will be returned.
E.g. Calling `step()` will return the `step` value, and calling `value([0, 0.5])` will set the lower and upper offsets to `0` and `0.5` respectively.

#### `disabled()`, `rangeSlideDisabled()`

The default parameter is set to `true`. So, if they are called without a parameter, they will set the corresponding values to `true`.
Thus, calling `disabled()` or `disabled(true)` will set `options.disabled = true` and calling `disabled(false)` will set `options.disabled = false`.

#### `thumbsDisabled()`

The default parameter is set to `[true, true]`. So, if it is called without a parameter, it will disable both thumbs. Example uses:

```js
//                          thumbs -> lower     upper
//                                    -----     -----
thumbsDisabled()                // disabled  disabled
thumbsDisabled(true)            // disabled  disabled
thumbsDisabled(false)           //  enabled   enabled
thumbsDisabled([])              //  enabled   enabled
thumbsDisabled([false])         //  enabled   enabled
thumbsDisabled([true])          // disabled   enabled
thumbsDisabled([true, false])   // disabled   enabled
thumbsDisabled([false, true])   //  enabled  disabled
thumbsDisabled([false, false])  //  enabled   enabled
thumbsDisabled([true, true])    // disabled  disabled
```

#### `currentValueIndex()`

Returns the index (`0` for the lower value and `1` for the upper value) of the value which is currently being modified.
Returns `-1` when the slider is idle.

#### `removeGlobalEventListeners()`

Removes the global event listeners. It should be called when removing the range slider element from the DOM dynamically.

## Elements

```html
<div class="range-slider"><!-- range slider element -->
    <input type="range" /><!-- hidden -->
    <input type="range" /><!-- hidden -->
    <div class="range-slider__thumb" data-lower></div>
    <div class="range-slider__thumb" data-upper></div>
    <div class="range-slider__range"></div>
</div>
```

`<div class="range-slider"></div>` is the wrapper element that was used to instantiate the range slider initially and is added with a CSS class named `range-slider`.

`<input type="range" />` elements are used to set values and are hidden.

`<div class="range-slider__thumb"></div>` elements are the slidable thumbs replacing the original thumbs from the `<input type="range" />` elements.

`<div class="range-slider__range"></div>` element fills up the space between the thumbs.


## Styling

**[View styled examples](https://n3r4zzurr0.in/range-slider-input/examples/#styled)**

```css
element-selector {
    /* CSS for the wrapper element */
}
element-selector[data-disabled] {
    /* CSS for disabled range slider element */
}
element-selector .range-slider__range {
    /* CSS for range */
}
element-selector .range-slider__range[data-active] {
    /* CSS for active (actively being dragged) range */
}
element-selector .range-slider__thumb {
    /* CSS for thumbs */
}
element-selector .range-slider__thumb[data-lower] {
    /* CSS for lower thumb */
}
element-selector .range-slider__thumb[data-upper] {
    /* CSS for upper thumb */
}
element-selector .range-slider__thumb[data-active] {
    /* CSS for active (actively being dragged) thumbs */
}
element-selector .range-slider__thumb[data-disabled] {
    /* CSS for disabled thumbs */
}
```

Refer to the `style.css` file to know more about styling the range slider element and its children.


## Component Wrappers

React: [react-range-slider-input](https://github.com/n3r4zzurr0/react-range-slider-input)


## License

MIT © [Utkarsh Verma](https://github.com/n3r4zzurr0)


CHANGELOG.md
### v2.4.x
Modified a property: [`onInput`](https://github.com/n3r4zzurr0/range-slider-input#options)

### v2.3.x
Added a return value property: [`removeGlobalEventListeners()`](https://github.com/n3r4zzurr0/range-slider-input#removeglobaleventlisteners)

### v2.2.x
Added a return value property: [`currentValueIndex()`](https://github.com/n3r4zzurr0/range-slider-input#currentvalueindex)

### v2.1.x
Added properties: `onThumbDragStart`, `onThumbDragEnd`, `onRangeDragStart` and `onRangeDragEnd`


### v2.0.x
Some new features have been introduced and some of the old features have been modified.

What's new?
- Keyboard accessibility
- ARIA value attributes
- Support for negative values
- Improved vertical orientation

What's changed?
- `<range>` element is replaced by `<div class="range-slider__range"></div>`
- `<thumb>` elements are replaced by `<div class="range-slider__thumb"></div>`
- `data-focused` is renamed to `data-active`.

### v1.0.x

Initial releases.


types/index.d.ts
type Orientation = 'vertical' | 'horizontal';
type Step = number | 'any';

export default function rangeSlider(element: Element, config?: RangeSliderConfig): RangeSlider;

export class RangeSlider {
    min(): number;
    min(min: number): void;

    max(): number;
    max(max: number): void;

    step(): Step;
    step(step: Step): void;

    value(): [number, number];
    value(value: [number, number]): void;

    orientation(): Orientation;
    orientation(orientation: Orientation): void;

    /** @param [disabled=true] */
    disabled(disabled?: boolean): void;

    /** @param [disabled=true] */
    rangeSlideDisabled(disabled?: boolean): void;

    /** @param [disabled=[true, true]] */
    thumbsDisabled(disabled?: [boolean, boolean]): void;

    currentValueIndex(): -1 | 0 | 1;

    removeGlobalEventListeners(): void;
}

export interface RangeSliderConfig {
    /** @default 0 */
    min?: number,
    /** @default 100 */
    max?: number,
    /** @default 1 */
    step?: Step,
    /** @default [25,75] */
    value?: [number, number],
    onInput?: (value: [number, number], userInteraction: boolean) => void,
    onThumbDragStart?: () => void,
    onThumbDragEnd?: () => void,
    onRangeDragStart?: () => void,
    onRangeDragEnd?: () => void,
    /** @default false */
    disabled?: boolean,
    /** @default false */
    rangeSlideDisabled?: boolean,
    /** @default [false, false] */
    thumbsDisabled?: [boolean, boolean],
    /** @default 'horizontal' */
    orientation?: Orientation,
}


src/index.style.js
module.exports = (element, options = {}) => {
  const isNumber = n => {
    // check for NaN explicitly
    // because with NaN, the second exp. evaluates to true
    return !isNaN(n) && +n + '' === n + ''
  }

  const setMinMaxProps = (min = 0, max = 0) => {
    return { min, max }
  }

  const iterateMinMaxProps = fn => {
    [MIN, MAX].forEach(fn)
  }

  const getSetProps = (condition, expression, fn) => {
    if (condition) { return expression } else { fn() }
  }

  const setNodeAttribute = (node, attribute, value = '') => {
    node.setAttribute(attribute, value)
  }

  const removeNodeAttribute = (node, attribute) => {
    node.removeAttribute(attribute)
  }

  const addNodeEventListener = (node, event, fn, isPointerEvent = true) => {
    // with options for pointer events
    node.addEventListener(event, fn, isPointerEvent ? { passive: false, capture: true } : {})
  }

  const removeNodeEventListener = (node, event, fn, isPointerEvent = true) => {
    // with options for pointer events
    node.removeEventListener(event, fn, isPointerEvent ? { passive: false, capture: true } : {})
  }

  const fallbackToDefault = (property, defaultValue) => {
    options[property] = {}.hasOwnProperty.call(options, property) ? options[property] : defaultValue
  }

  const ifVerticalElse = (vertical, horizontal) => {
    return options.orientation === VERTICAL ? vertical : horizontal
  }

  const currentIndex = i => {
    return i === 1 ? index.max : index.min
  }

  // Set min and max values to 1 (arbitrarily) if any of the min or max values are "invalid"
  // Setting both values 1 will disable the slider
  // Called when,
  // -> the element is initially set
  // -> min or max properties are modified
  const safeMinMaxValues = () => {
    let error = false
    if (!isNumber(options.min) || !isNumber(options.max)) { error = true }
    options.min = error ? 1 : +options.min
    options.max = error ? 1 : +options.max
  }

  // Reframe the thumbsDisabled value if "invalid"
  // Called when,
  // -> the element is initially set
  // -> thumbsDisabled property is modified
  const safeThumbsDisabledValues = () => {
    if (options.thumbsDisabled instanceof Array) {
      if (options.thumbsDisabled.length === 1) { options.thumbsDisabled.push(false) }
      if (options.thumbsDisabled.length !== 1 && options.thumbsDisabled.length !== 2) { options.thumbsDisabled = [false, false] }
    } else { options.thumbsDisabled = [options.thumbsDisabled, options.thumbsDisabled] }

    // Boolean Values
    options.thumbsDisabled[0] = !!options.thumbsDisabled[0]
    options.thumbsDisabled[1] = !!options.thumbsDisabled[1]
  }

  // Called when,
  // -> the element is initially set
  // -> min, max, step or value properties are modified
  // -> thumbs are dragged
  // -> element is clicked upon
  // -> an arrow key is pressed
  const setValue = (newValue, forceSet = false, callback = true, userInteraction = true) => {
    // Current value as set in the input elements
    // which could change while changing min, max and step values
    const currentValue = setMinMaxProps(input[index.min].value, input[index.max].value)

    // var value is synced with the values set in the input elements if no newValue is passed
    newValue = newValue || currentValue

    input[index.min].value = newValue.min
    input[index.max].value = (thumbDrag || forceSet) ? newValue.max : (newValue.min + rangeWidth)
    syncValues()

    // Check if the thumbs cross each other
    if (value.min > value.max) {
      // Switch thumb indexes
      index.min = +!index.min
      index.max = +!index.max

      // Switch thumb attributes
      removeNodeAttribute(thumb[index.min], DATA_UPPER)
      removeNodeAttribute(thumb[index.max], DATA_LOWER)
      setNodeAttribute(thumb[index.min], DATA_LOWER)
      setNodeAttribute(thumb[index.max], DATA_UPPER)

      // Switch thumb drag labels
      if (thumbDrag) { thumbDrag = thumbDrag === MIN ? MAX : MIN }

      syncValues()
    }

    sliderValue = forceSet ? value : newValue

    let valueSet = false

    if (currentValue.min !== input[index.min].value || forceSet) { valueSet = true }

    if (currentValue.max !== input[index.max].value || forceSet) { valueSet = true }

    // Update the positions, dimensions and aria attributes everytime a value is set
    // and call the onInput function from options (if set)
    if (valueSet) {
      if (callback && options.onInput) { options.onInput([value.min, value.max], userInteraction) }
      syncThumbDimensions()
      updateThumbs()
      updateRange()
      updateAriaValueAttributes()
    }
  }

  // Sync var value with the input elements
  const syncValues = () => {
    iterateMinMaxProps(_ => {
      value[_] = +input[index[_]].value
    })
  }

  // Called when,
  // -> setValue is called and a value is set
  // -> window is resized
  const updateThumbs = () => {
    iterateMinMaxProps(_ => {
      thumb[index[_]].style[ifVerticalElse('top', 'left')] = `calc(${((value[_] - options.min) / maxRangeWidth) * 100}% + ${(0.5 - ((value[_] - options.min) / maxRangeWidth)) * ifVerticalElse(thumbHeight, thumbWidth)[_]}px)`
    })
  }

  // Called when,
  // -> setValue is called and a value is set
  // -> window is resized
  const updateRange = () => {
    const deltaOffset = ((0.5 - ((value.min - options.min) / maxRangeWidth)) * ifVerticalElse(thumbHeight, thumbWidth).min) / element[`client${ifVerticalElse('Height', 'Width')}`]
    const deltaDimension = ((0.5 - ((value.max - options.min) / maxRangeWidth)) * ifVerticalElse(thumbHeight, thumbWidth).max) / element[`client${ifVerticalElse('Height', 'Width')}`]
    range.style[ifVerticalElse('top', 'left')] = `${(((value.min - options.min) / maxRangeWidth) + deltaOffset) * 100}%`
    range.style[ifVerticalElse('height', 'width')] = `${(((value.max - options.min) / maxRangeWidth) - ((value.min - options.min) / maxRangeWidth) - deltaOffset + deltaDimension) * 100}%`
  }

  // Called when,
  // -> the element is initially set
  // -> min, max or value properties are modified
  // -> range is dragged
  // -> thumbs are disabled / enabled
  const updateRangeLimits = () => {
    iterateMinMaxProps((_, i) => {
      rangeLimits[_] = options.thumbsDisabled[i] ? value[_] : options[_]
    })
  }

  // Called when,
  // -> thumbs are initially set
  // -> thumbs are disabled / enabled
  const updateTabIndexes = () => {
    iterateMinMaxProps((_, i) => {
      if (!options.disabled && !options.thumbsDisabled[i]) { setNodeAttribute(thumb[currentIndex(i)], TABINDEX, 0) } else { removeNodeAttribute(thumb[currentIndex(i)], TABINDEX) }
    })
  }

  // Called when,
  // -> setValue is called and a value is set
  const updateAriaValueAttributes = () => {
    iterateMinMaxProps(_ => {
      setNodeAttribute(thumb[index[_]], 'aria-valuemin', options.min)
      setNodeAttribute(thumb[index[_]], 'aria-valuemax', options.max)
      setNodeAttribute(thumb[index[_]], 'aria-valuenow', value[_])
      setNodeAttribute(thumb[index[_]], 'aria-valuetext', value[_])
    })
  }

  // Called when,
  // -> disabled property is modified
  const updateDisabledState = () => {
    if (options.disabled) { setNodeAttribute(element, DATA_DISABLED) } else { removeNodeAttribute(element, DATA_DISABLED) }
  }

  // Called when,
  // -> thumbsDisabled property is modified
  const updateThumbsDisabledState = () => {
    options.thumbsDisabled.forEach((d, i) => {
      const currIndex = currentIndex(i)
      if (d) {
        setNodeAttribute(thumb[currIndex], DATA_DISABLED)
        setNodeAttribute(thumb[currIndex], 'aria-disabled', true)
      } else {
        removeNodeAttribute(thumb[currIndex], DATA_DISABLED)
        setNodeAttribute(thumb[currIndex], 'aria-disabled', false)
      }
    })
  }

  // Called when,
  // -> min or max values are modified
  const updateLimits = (limit, m = false) => {
    options[limit] = m
    safeMinMaxValues()
    iterateMinMaxProps(_ => {
      input[0][_] = options[_]
      input[1][_] = options[_]
    })
    maxRangeWidth = options.max - options.min
    setValue('', true, true, false)
    updateRangeLimits()
  }

  // Called when,
  // -> the element is initially set
  // -> orientation property is modified
  const updateOrientation = () => {
    if (options.orientation === VERTICAL) { setNodeAttribute(element, DATA_VERTICAL) } else { removeNodeAttribute(element, DATA_VERTICAL) }
    range.style[ifVerticalElse('left', 'top')] = ''
    range.style[ifVerticalElse('width', 'height')] = ''
    thumb[0].style[ifVerticalElse('left', 'top')] = ''
    thumb[1].style[ifVerticalElse('left', 'top')] = ''
  }

  // thumb width & height values are to be synced with the CSS values for correct calculation of
  // thumb position and range width & position
  // Called when,
  // -> setValue is called and a value is set (called before updateThumbs() and updateRange())
  // -> thumb / range drag is initiated
  // -> window is resized
  const syncThumbDimensions = () => {
    iterateMinMaxProps(_ => {
      thumbWidth[_] = float(style(thumb[index[_]]).width)
      thumbHeight[_] = float(style(thumb[index[_]]).height)
    })
  }

  // thumb position calculation depending upon the pointer position
  const currentPosition = (e, node) => {
    const currPos = ((node[`offset${ifVerticalElse('Top', 'Left')}`] + (e[`client${ifVerticalElse('Y', 'X')}`] - node.getBoundingClientRect()[ifVerticalElse('top', 'left')]) - (thumbDrag ? ((0.5 - (value[thumbDrag] - options.min) / maxRangeWidth) * ifVerticalElse(thumbHeight, thumbWidth)[thumbDrag]) : 0)) / element[`client${ifVerticalElse('Height', 'Width')}`]) * maxRangeWidth + options.min
    if (currPos < options.min) { return options.min }
    if (currPos > options.max) { return options.max }
    return currPos
  }

  const doesntHaveClassName = (e, className) => {
    return !e.target.classList.contains(className)
  }

  const elementFocused = e => {
    let setFocus = false

    if (!options.disabled && ((doesntHaveClassName(e, 'range-slider__thumb') && doesntHaveClassName(e, 'range-slider__range')) || (options.rangeSlideDisabled && doesntHaveClassName(e, 'range-slider__thumb')))) { setFocus = true }

    // No action if both thumbs are disabled
    if (setFocus && options.thumbsDisabled[0] && options.thumbsDisabled[1]) { setFocus = false }

    if (setFocus) {
      const currPos = currentPosition(e, range)
      const deltaMin = abs(value.min - currPos)
      const deltaMax = abs(value.max - currPos)

      if (options.thumbsDisabled[0]) {
        if (currPos >= value.min) {
          setValue(setMinMaxProps(value.min, currPos), true)
          initiateThumbDrag(e, index.max, thumb[index.max])
        }
      } else if (options.thumbsDisabled[1]) {
        if (currPos <= value.max) {
          setValue(setMinMaxProps(currPos, value.max), true)
          initiateThumbDrag(e, index.min, thumb[index.min])
        }
      } else {
        let nearestThumbIndex = index.max
        if (deltaMin === deltaMax) { setValue(setMinMaxProps(value.min, currPos), true) } else {
          setValue(setMinMaxProps(deltaMin < deltaMax ? currPos : value.min, deltaMax < deltaMin ? currPos : value.max), true)
          nearestThumbIndex = deltaMin < deltaMax ? index.min : index.max
        }
        initiateThumbDrag(e, nearestThumbIndex, thumb[nearestThumbIndex])
      }
    }
  }

  const initiateDrag = (e, node) => {
    syncThumbDimensions()
    setNodeAttribute(node, DATA_ACTIVE)
    startPos = currentPosition(e, node)
    isDragging = true
  }

  const initiateThumbDrag = (e, i, node) => {
    if (!options.disabled && !options.thumbsDisabled[currentIndex(i)]) {
      initiateDrag(e, node)
      thumbDrag = index.min === i ? MIN : MAX
      if (options.onThumbDragStart) { options.onThumbDragStart() }
    }
  }

  const initiateRangeDrag = e => {
    if (!options.disabled && !options.rangeSlideDisabled) {
      initiateDrag(e, range)
      rangeWidth = value.max - value.min
      thumbDrag = false
      if (options.onRangeDragStart) { options.onRangeDragStart() }
    }
  }

  const drag = e => {
    if (isDragging) {
      const lastPos = currentPosition(e, range)
      const delta = lastPos - startPos

      let min = value.min
      let max = value.max
      const lower = thumbDrag ? rangeLimits.min : options.min
      const upper = thumbDrag ? rangeLimits.max : options.max

      if (!thumbDrag || thumbDrag === MIN) { min = thumbDrag ? lastPos : (sliderValue.min + delta) }
      if (!thumbDrag || thumbDrag === MAX) { max = thumbDrag ? lastPos : (sliderValue.max + delta) }

      if (min >= lower && min <= upper && max >= lower && max <= upper) {
        setValue({ min, max })
        startPos = lastPos
      } else {
        // When min thumb reaches upper limit
        if (min > upper && thumbDrag) {
          setValue(setMinMaxProps(upper, upper))
          startPos = lastPos
        }
        // When max thumb reaches lower limit
        if (max < lower && thumbDrag) {
          setValue(setMinMaxProps(lower, lower))
          startPos = lastPos
        }
        // When range / min thumb reaches lower limit
        if (min < lower) {
          if (!thumbDrag) { setValue(setMinMaxProps(lower, value.max - value.min + lower)) } else { setValue(setMinMaxProps(lower, value.max)) }
          startPos = lastPos
        }
        // When range / max thumb reaches upper limit
        if (max > upper) {
          if (!thumbDrag) { setValue(setMinMaxProps(value.min - value.max + upper, upper)) } else { setValue(setMinMaxProps(value.min, upper)) }
          startPos = lastPos
        }
      }
      if (!thumbDrag) { updateRangeLimits() }
    }
  }

  const unfocus = () => {
    if (isDragging) {
      removeNodeAttribute(thumb[0], DATA_ACTIVE)
      removeNodeAttribute(thumb[1], DATA_ACTIVE)
      removeNodeAttribute(range, DATA_ACTIVE)
      isDragging = false
      if (thumbDrag) {
        if (options.onThumbDragEnd) { options.onThumbDragEnd() }
      } else {
        if (options.onRangeDragEnd) { options.onRangeDragEnd() }
      }
    }
  }

  const resize = () => {
    syncThumbDimensions()
    updateThumbs()
    updateRange()
  }

  const actualStepValue = () => {
    const step = float(input[0].step)
    return input[0].step === ANY ? ANY : ((step === 0 || isNaN(step)) ? 1 : step)
  }

  // Step value (up or down) using arrow keys
  const stepValue = (i, key) => {
    const direction = (key === 37 || key === 40 ? -1 : 1) * ifVerticalElse(-1, 1)

    if (!options.disabled && !options.thumbsDisabled[currentIndex(i)]) {
      let step = actualStepValue()
      step = step === ANY ? 1 : step

      let min = value.min + step * (index.min === i ? direction : 0)
      let max = value.max + step * (index.max === i ? direction : 0)

      // When min thumb reaches upper limit
      if (min > rangeLimits.max) { min = rangeLimits.max }

      // When max thumb reaches lower limit
      if (max < rangeLimits.min) { max = rangeLimits.min }

      setValue({ min, max }, true)
    }
  }

  // Aliases
  const abs = Math.abs
  const float = parseFloat
  const style = window.getComputedStyle

  // Values
  const MIN = 'min'
  const MAX = 'max'
  const ANY = 'any'
  const VERTICAL = 'vertical'
  const TABINDEX = 'tabindex'

  // Data Attributes
  const DATA_LOWER = 'data-lower'
  const DATA_UPPER = 'data-upper'
  const DATA_ACTIVE = 'data-active'
  const DATA_VERTICAL = 'data-vertical'
  const DATA_DISABLED = 'data-disabled'

  // Actual value
  const value = setMinMaxProps()

  // Thumb indexes for min and max values
  // (swapped when the thumbs cross each other)
  const index = setMinMaxProps(0, 1)

  // Thumb width & height for calculation of exact positions and sizes of horizontal thumbs and range
  const thumbWidth = setMinMaxProps()
  const thumbHeight = setMinMaxProps()

  // Slidable range limits (when a thumb is dragged)
  const rangeLimits = setMinMaxProps()

  // Slider value depending on the user interaction
  let sliderValue = setMinMaxProps()

  // For dragging thumbs and range
  let maxRangeWidth = 0
  let rangeWidth = 0
  let isDragging = false
  let thumbDrag = false
  let startPos = 0

  if (!this.cssInjected) {
    // Inject Core CSS (on first call)
    const style = document.createElement('style')
    style.textContent = '.range-slider{touch-action:none;-webkit-tap-highlight-color:transparent;-webkit-user-select:none;user-select:none;cursor:pointer;display:block;position:relative;width:100%;height:8px;background:#ddd;border-radius:4px}.range-slider[data-vertical]{height:100%;width:8px}.range-slider[data-disabled]{opacity:.5;cursor:not-allowed}.range-slider .range-slider__thumb{position:absolute;z-index:3;top:50%;width:24px;height:24px;transform:translate(-50%,-50%);border-radius:50%;background:#2196f3}.range-slider .range-slider__thumb:focus-visible{outline:0;box-shadow:0 0 0 6px rgba(33,150,243,.5)}.range-slider[data-vertical] .range-slider__thumb{left:50%}.range-slider .range-slider__thumb[data-disabled]{z-index:2}.range-slider .range-slider__range{position:absolute;z-index:1;transform:translate(0,-50%);top:50%;width:100%;height:100%;background:#51adf6}.range-slider[data-vertical] .range-slider__range{left:50%;transform:translate(-50%,0)}.range-slider input[type=range]{-webkit-appearance:none;pointer-events:none;position:absolute;z-index:2;top:0;left:0;width:0;height:0;background-color:transparent}.range-slider input[type=range]::-webkit-slider-thumb{-webkit-appearance:none;appearance:none}.range-slider input[type=range]::-moz-range-thumb{width:0;height:0;border:0}.range-slider input[type=range]:focus{outline:0}'
    const head = document.head
    const tag = document.querySelector('head>style,head>link')
    if (tag) { head.insertBefore(style, tag) } else { head.appendChild(style) }
    this.cssInjected = true
  }

  // Set options to default values if not set
  fallbackToDefault('rangeSlideDisabled', false)
  fallbackToDefault('thumbsDisabled', [false, false])
  fallbackToDefault('orientation', 'horizontal')
  fallbackToDefault('disabled', false)
  fallbackToDefault('onThumbDragStart', false)
  fallbackToDefault('onRangeDragStart', false)
  fallbackToDefault('onThumbDragEnd', false)
  fallbackToDefault('onRangeDragEnd', false)
  fallbackToDefault('onInput', false)
  fallbackToDefault('value', [25, 75])
  fallbackToDefault('step', 1)
  fallbackToDefault('min', 0)
  fallbackToDefault('max', 100)

  safeMinMaxValues()
  safeThumbsDisabledValues()

  // Fill wrapper element
  element.innerHTML = `<input type="range" min="${options.min}" max="${options.max}" step="${options.step}" value="${options.value[0]}" disabled><input type="range" min="${options.min}" max="${options.max}" step="${options.step}" value="${options.value[1]}" disabled><div role="slider" class="range-slider__thumb" ${DATA_LOWER}></div><div role="slider" class="range-slider__thumb" ${DATA_UPPER}></div><div class="range-slider__range"></div>`
  element.classList.add('range-slider')

  const range = element.querySelector('.range-slider__range')
  const input = element.querySelectorAll('input')
  const thumb = element.querySelectorAll('.range-slider__thumb')

  // Set initial values
  maxRangeWidth = options.max - options.min
  setValue('', true, false)
  updateRangeLimits()
  updateDisabledState()
  updateThumbsDisabledState()
  updateTabIndexes()
  updateOrientation()

  // Add listeners to element
  addNodeEventListener(element, 'pointerdown', e => { elementFocused(e) })

  // Add listeners to thumbs and set [data-disabled] on disabled thumbs
  Array.from(thumb).forEach((t, i) => {
    addNodeEventListener(t, 'pointerdown', e => { initiateThumbDrag(e, i, t) })
    addNodeEventListener(t, 'keydown', e => {
      if (e.which >= 37 && e.which <= 40) {
        e.preventDefault()
        stepValue(i, e.which)
      }
    })
  })

  // Add listeners to range
  addNodeEventListener(range, 'pointerdown', e => { initiateRangeDrag(e) })

  // Add global listeners
  addNodeEventListener(document, 'pointermove', drag)
  addNodeEventListener(document, 'pointerup', unfocus)
  addNodeEventListener(window, 'resize', resize)

  return {
    min: (m = false) => {
      return getSetProps(!m && m !== 0, options.min, () => {
        updateLimits(MIN, m)
      })
    },
    max: (m = false) => {
      return getSetProps(!m && m !== 0, options.max, () => {
        updateLimits(MAX, m)
      })
    },
    step: (s = false) => {
      return getSetProps(!s, actualStepValue(), () => {
        input[0].step = s
        input[1].step = s
        setValue('', true, true, false)
      })
    },
    value: (v = false) => {
      return getSetProps(!v, [value.min, value.max], () => {
        setValue(setMinMaxProps(v[0], v[1]), true, true, false)
        updateRangeLimits()
      })
    },
    orientation: (o = false) => {
      return getSetProps(!o, options.orientation, () => {
        options.orientation = o
        updateOrientation()
        setValue('', true, true, false)
      })
    },
    disabled: (d = true) => {
      options.disabled = !!d
      updateDisabledState()
    },
    thumbsDisabled: (t = [true, true]) => {
      options.thumbsDisabled = t
      safeThumbsDisabledValues()
      updateRangeLimits()
      updateTabIndexes()
      updateThumbsDisabledState()
    },
    rangeSlideDisabled: (d = true) => {
      options.rangeSlideDisabled = !!d
    },
    currentValueIndex: () => {
      return thumbDrag ? (thumbDrag === MIN ? 0 : 1) : -1
    },
    removeGlobalEventListeners: () => {
      removeNodeEventListener(document, 'pointermove', drag)
      removeNodeEventListener(document, 'pointerup', unfocus)
      removeNodeEventListener(window, 'resize', resize)
    }
  }
}
